### PR TITLE
Add key revocation option

### LINKED ISSUE #465: revoking a public key
Is this possible? I see in openpgp.js

> // TODO
> Key.prototype.revoke = function () {};

so I guess not, 
would be great though if we could create a revocation certificate so a public key could be revoked.

### PR DESCRIPTION
**Pull request by MaximilianKrambach**
Fixes #465 

Added a method for Keys to be revoked. It is still not fully merge-ready, as I don't yet fully understand how to handle errors in openpgpjs; I'd appreciate some pointers.

For subkey revocations I have a similar method prepared, but I don't fully grasp get what constitutes a 'subkey'. I'll be reading it up.

### CODE REVIEW

File: src/key.js [hunk 1]
```diff
@@ -623,9 +625,38 @@ async function mergeSignatures(source, dest, attr, checkFn) {
...
-// TODO
-Key.prototype.revoke = function() {
-
+/**
+ * Revokes a Key
+ * @param {Integer} reasonFlag optional reason for revocation
+ * @param {Integer} reasonString optional additional description string
+ * @async
+ */
+Key.prototype.revoke = async function(reasonFlag, reasonString){
```
Comment (snowyu) [inline src/key.js]: The reasonString should be used more than reasonFlag, so exchange the params position may be good.

```js
Key.prototype.revoke = async function(reasonString, reasonFlag)
```

File: src/key.js [hunk 2]
```diff
@@ -623,9 +625,38 @@ async function mergeSignatures(source, dest, attr, checkFn) {
...
+  }
+  if (!Number.isInteger(reasonFlag) || reasonFlag < 0 || reasonFlag > 110) {
+    return false;
+    // TODO error- reason flag is not valid
+  }
+  if (reasonString !== undefined && !util.isString(reasonString)){
+    // TODO error- reason is not valid
+    return false;
+  }
+  if (Key.isPrivate !== true) {
```
Comment (snowyu) [inline src/key.js]: It should be:
```js
if (!this.isPrivate()) {
```

File: test/general/key.js [hunk 3]
```diff
@@ -1737,4 +1737,22 @@ p92yZgB3r2+f6/GIe2+7
...
+    const opt = {
+      userIds: { name: 'Test User', email: 'text@example.com' },
+      passphrase: 'secret'
+    };
+    return openpgp.generateKey(opt).then(function(newKey) {
+      expect(newKey.key).to.exist;
+      newKey.key.isRevoked().then(function(answer){
+        expect(answer).to.be.false;
+        newKey.key.revoke().then(function(){
+          newKey.key.isRevoked().then(function(answer2){
```
Comment (snowyu) [inline test/general/key.js]: Wrong Promise!

```js
 return openpgp.generateKey(opt).then(function(newKey) {
      expect(newKey.key).to.exist;
      return newKey.key.isRevoked().then(function(answer){
        expect(answer).to.be.false;
        return newKey.key.revoke().then(function(){
          return newKey.key.isRevoked().then(function(answer2){
```
More clear:

```js
    return openpgp.generateKey(opt).then(function(newKey) {
      expect(newKey.key).to.exist;
    }).then(function(){
      return newKey.key.isRevoked();  
    }).then(function(answer){
      expect(answer).to.be.false;
      return newKey.key.revoke();
    }).then(function(){
      return newKey.key.isRevoked();
    }).then(function(answer){
      expect(answer).to.be.true;
    });
```

File: src/key.js [hunk 4]
```diff
@@ -623,9 +625,38 @@ async function mergeSignatures(source, dest, attr, checkFn) {
...
+    // TODO can't revoke if we don't have the private Key
+    return false;
+  }
+  const revSig = new Signature();
```
Comment (snowyu) [inline src/key.js]: Wrong Usage

**Comment by MaximilianKrambach** [comment]
@snowyu I suppose I can abandon working on this due to #730 implementing the same?

**Comment by snowyu** [comment]
Yes.

**Pull request closed without merging**